File: python/npgen2007/variant.py
# Copyright (C) 2007 Time Intermedia Corporation <puzzle@timedia.co.jp>
# Derived from NPGenerator V2.0.2. SPDX-License-Identifier: GPL-3.0-or-later
"""Variant constraint construction and the original random block splitter."""

from __future__ import annotations

from dataclasses import dataclass, field
from math import isqrt

from .gridio import read_block_array
from .model import BlockConstraint
from .random import JavaRandom

DEFAULT_SIZE = 9
MIN_SIZE = 2
MAX_SIZE = 25


@dataclass(frozen=True)
class Variant:
    size: int
    block: BlockConstraint
    block_array: tuple[int, ...]
    diagonal: bool
    default_block: bool
    vertical: bool = True
    horizontal: bool = True


@dataclass
class ProblemBuilder:
    """Python counterpart of the original ProblemBuilder assembly order."""

    size: int
    vertical: bool = True
    horizontal: bool = True
    diagonal: bool = False
    default_block: bool = True
    rectangle_width: int = -1
    rectangle_height: int = -1
    groups: list[list[int]] = field(default_factory=list)

    def build(self, diagonal_last: bool = False) -> list[list[int]]:
        blocks: list[list[int]] = []
        if self.vertical:
            _add_vertical(blocks, self.size)
        if self.horizontal:
            _add_horizontal(blocks, self.size)
        if self.diagonal and not diagonal_last:
            _add_diagonal(blocks, self.size)
        if self.default_block:
            square = isqrt(self.size)
            _add_rectangle(blocks, self.size, square, square)
        elif self.rectangle_width > 0 and self.rectangle_height > 0:
            _add_rectangle(
                blocks,
                self.size,
                self.rectangle_width,
                self.rectangle_height,
            )
        for labels in self.groups:
            _add_by_array(blocks, labels, self.size)
        if self.diagonal and diagonal_last:
            _add_diagonal(blocks, self.size)
        return blocks


def _add_vertical(blocks: list[list[int]], size: int) -> None:
    for column in range(size):
        blocks.append([row * size + column for row in range(size)])


def _add_horizontal(blocks: list[list[int]], size: int) -> None:
    for row in range(size):
        blocks.append([row * size + column for column in range(size)])


def _add_diagonal(blocks: list[list[int]], size: int) -> None:
    blocks.append([index * size + index for index in range(size)])
    blocks.append([
        (size - 1 - index) * size + index for index in range(size)
    ])


def _add_rectangle(
    blocks: list[list[int]], size: int, width: int, height: int
) -> None:
    for left in range(0, size, width):
        for top in range(0, size, height):
            blocks.append([
                (top + y) * size + left + x
                for x in range(width)
                for y in range(height)
            ])


def _add_by_array(
    blocks: list[list[int]], labels: list[int], size: int
) -> None:
    label_to_index: dict[int, int] = {}
    for label in labels:
        if label != 0 and label not in label_to_index:
            label_to_index[label] = len(label_to_index)
    groups = [[] for _ in label_to_index]
    for cell, label in enumerate(labels):
        if label != 0:
            groups[label_to_index[label]].append(cell)
    if any(len(group) != size for group in groups):
        raise ValueError(
            f"every block must contain exactly {size} cells"
        )
    blocks.extend(groups)


class _DisjointSet:
    def __init__(self, size: int) -> None:
        self.parent = list(range(size))

    def find(self, value: int) -> int:
        parent = self.parent[value]
        if parent == value:
            return value
        root = self.find(parent)
        self.parent[value] = root
        return root

    def union(self, first: int, second: int) -> None:
        self.parent[self.find(first)] = self.find(second)

    def is_same_group(self, first: int, second: int) -> bool:
        return self.find(first) == self.find(second)


class BlockSplit:
    """Exact port of core/BlockSplit, including its traversal quirk."""

    def __init__(
        self, width: int, height: int, random: JavaRandom
    ) -> None:
        self.width = width
        self.height = height
        self.random = random
        self.vertical_wall: list[list[bool]] = []
        self.horizontal_wall: list[list[bool]] = []

    @staticmethod
    def _index(x: int, y: int, width: int) -> int:
        return y * width + x

    def _generate_maze(self, width: int, height: int) -> None:
        self.vertical_wall = [
            [False] * (height + 1) for _ in range(width + 1)
        ]
        self.horizontal_wall = [
            [False] * (height + 1) for _ in range(width + 1)
        ]
        groups = _DisjointSet(width * height)
        edges: list[tuple[int, int]] = []
        for x in range(width):
            for y in range(height):
                self.vertical_wall[x][y] = True
                self.horizontal_wall[x][y] = True
                cell = self._index(x, y, width)
                if x + 1 < width:
                    edges.append((cell, self._index(x + 1, y, width)))
                if y + 1 < height:
                    edges.append((cell, self._index(x, y + 1, width)))
        self.random.shuffle(edges)
        remaining = width * height - 1
        for first, second in edges:
            if remaining <= 0:
                break
            first_x, first_y = first % width, first // width
            second_x, second_y = second % width, second // width
            if groups.is_same_group(first, second):
                continue
            groups.union(first, second)
            if first_x == second_x:
                self.vertical_wall[first_x][first_y] = False
            elif first_y == second_y:
                self.horizontal_wall[first_x][first_y] = False
            remaining -= 1

    def _is_movable(
        self, px: int, py: int, qx: int, qy: int
    ) -> bool:
        if px == qx:
            if py > qy:
                return self._is_movable(px, qy, qx, py)
            if py + 1 != qy:
                return False
            if py & 1 == 0:
                return True
            return not self.vertical_wall[px // 2][py // 2]
        if py == qy:
            if px > qx:
                return self._is_movable(qx, py, px, qy)
            if px + 1 != qx:
                return False
            if px & 1 == 0:
                return True
            return not self.horizontal_wall[px // 2][py // 2]
        return False

    def _walk_maze(self) -> list[int]:
        dx = (0, 1, 0, -1)
        dy = (1, 0, -1, 0)
        x = self.random.next_int(self.width // 2) * 2
        y = self.random.next_int(self.height // 2) * 2
        direction = 0
        result = [-1] * (self.width * self.height)
        block_id = 0
        count = 0
        yet = True
        while yet:
            yet = False
            result[self._index(x, y, self.width)] = block_id
            count += 1
            if count == self.width:
                block_id += 1
                count = 0
            for index in range(3, 7):
                next_direction = (index + direction) % 4
                next_x = x + dx[next_direction]
                next_y = y + dy[next_direction]
                if (
                    self.width % 2 != 0
                    and self.height % 2 != 0
                    and next_x == self.width - 1
                    and next_y == self.height - 1
                    and count != 0
                    and result[
                        self._index(next_x, next_x, self.width)
                    ] == -1
                ):
                    # The original intentionally uses next_x twice.
                    result[
                        self._index(next_x, next_x, self.width)
                    ] = block_id
                    count += 1
                    if count == self.width:
                        block_id += 1
                        count = 0
                if (
                    0 <= next_x < self.width
                    and 0 <= next_y < self.height
                    and self._is_movable(x, y, next_x, next_y)
                    and result[
                        self._index(next_x, next_y, self.width)
                    ] == -1
                ):
                    yet = True
                    direction = next_direction
                    x, y = next_x, next_y
                    break
        return result

    def split_block(self) -> list[int]:
        width = self.width // 2
        height = self.height // 2
        self._generate_maze(width, height)
        if self.width % 2 != 0:
            for index in range(height):
                self.vertical_wall[width][index] = True
                self.horizontal_wall[width - 1][index] = False
        if self.height % 2 != 0:
            for index in range(width):
                self.vertical_wall[index][height - 1] = False
                self.horizontal_wall[index][height] = True
        return self._walk_maze()


def rectangle_block_array(
    size: int, width: int, height: int
) -> list[int]:
    blocks_across = size // width
    return [
        (row // height) * blocks_across + column // width + 1
        for row in range(size)
        for column in range(size)
    ]


def _normalize_block_array(size: int, labels: list[int]) -> list[int]:
    if len(labels) != size * size:
        raise ValueError(
            f"block grid must contain exactly {size * size} cells"
        )
    normalized_labels: dict[int, int] = {}
    result: list[int] = []
    counts = [0] * size
    for label in labels:
        normalized = normalized_labels.get(label)
        if normalized is None:
            if len(normalized_labels) == size:
                raise ValueError(
                    f"block grid must contain exactly {size} blocks"
                )
            normalized = len(normalized_labels) + 1
            normalized_labels[label] = normalized
        result.append(normalized)
        counts[normalized - 1] += 1
    if len(normalized_labels) != size:
        raise ValueError(
            f"block grid must contain exactly {size} blocks"
        )
    if any(count != size for count in counts):
        raise ValueError(
            f"every block must contain exactly {size} cells"
        )
    return result


def _base_builder(
    size: int, vertical: bool, horizontal: bool, diagonal: bool
) -> ProblemBuilder:
    return ProblemBuilder(
        size=size,
        vertical=vertical,
        horizontal=horizontal,
        diagonal=diagonal,
    )


def _finish_variant(
    builder: ProblemBuilder,
    block_array: list[int],
    default_block: bool,
    xml_order: bool = False,
) -> Variant:
    return Variant(
        size=builder.size,
        block=BlockConstraint(builder.build(xml_order), builder.size),
        block_array=tuple(block_array),
        diagonal=builder.diagonal,
        default_block=default_block,
        vertical=builder.vertical,
        horizontal=builder.horizontal,
    )


def _build_array_variant(
    size: int,
    labels: list[int],
    diagonal: bool,
    vertical: bool = True,
    horizontal: bool = True,
    xml_order: bool = False,
) -> Variant:
    normalized = _normalize_block_array(size, labels)
    builder = _base_builder(size, vertical, horizontal, diagonal)
    builder.default_block = False
    builder.groups.append(normalized)
    return _finish_variant(builder, normalized, False, xml_order)


def build_variant(
    size: int,
    block_spec: str | None,
    diagonal: bool,
    random: JavaRandom,
    vertical: bool = True,
    horizontal: bool = True,
    xml_order: bool = False,
) -> Variant:
    require_size(size)
    if block_spec is None:
        square = isqrt(size)
        if square * square != size:
            raise ValueError(
                "--blocks is required when --size is not a perfect square"
            )
        builder = _base_builder(size, vertical, horizontal, diagonal)
        builder.default_block = True
        return _finish_variant(
            builder,
            rectangle_block_array(size, square, square),
            True,
            xml_order,
        )
    if block_spec == "random":
        return _build_array_variant(
            size,
            BlockSplit(size, size, random).split_block(),
            diagonal,
            vertical,
            horizontal,
            xml_order,
        )
    if block_spec.startswith("@"):
        if len(block_spec) == 1:
            raise ValueError("--blocks @file requires a file name")
        return _build_array_variant(
            size,
            read_block_array(block_spec[1:], size),
            diagonal,
            vertical,
            horizontal,
            xml_order,
        )
    dimensions = block_spec.lower().split("x")
    if len(dimensions) != 2:
        raise ValueError("--blocks must be WxH, random, or @file.txt")
    try:
        width, height = (int(value, 10) for value in dimensions)
    except ValueError as error:
        raise ValueError(
            "--blocks WxH requires integer dimensions"
        ) from error
    if (
        width <= 0
        or height <= 0
        or width * height != size
        or size % width != 0
        or size % height != 0
    ):
        raise ValueError("--blocks WxH requires W*H == size")
    builder = _base_builder(size, vertical, horizontal, diagonal)
    builder.default_block = False
    builder.rectangle_width = width
    builder.rectangle_height = height
    return _finish_variant(
        builder,
        rectangle_block_array(size, width, height),
        False,
        xml_order,
    )


def build_xml_variant(
    source: object,
    force_diagonal: bool,
    no_vertical: bool = False,
    no_horizontal: bool = False,
) -> Variant:
    size = getattr(source, "num_size")
    require_size(size)
    vertical = getattr(source, "vertical") and not no_vertical
    horizontal = getattr(source, "horizontal") and not no_horizontal
    diagonal = getattr(source, "diagonal") or force_diagonal
    builder = _base_builder(size, vertical, horizontal, diagonal)
    group_arrays = getattr(source, "group_arrays")
    if getattr(source, "default_block"):
        square = isqrt(size)
        if square * square != size:
            raise ValueError(
                "XML default-block requires a perfect-square size"
            )
        builder.default_block = True
        block_array = list(
            rectangle_block_array(size, square, square)
        )
    else:
        if not group_arrays:
            raise ValueError(
                "XML custom block constraint is missing <group>"
            )
        builder.default_block = False
        block_array = list(group_arrays[0])
    for labels in group_arrays:
        builder.groups.append(list(labels))
    return _finish_variant(
        builder,
        block_array,
        getattr(source, "default_block"),
        True,
    )


def require_size(size: int) -> None:
    if not MIN_SIZE <= size <= MAX_SIZE:
        raise ValueError("--size must be between 2 and 25")

File: typescript/src/variant.ts
/*
 * Copyright (C) 2007 Time Intermedia Corporation <puzzle@timedia.co.jp>
 * TypeScript rewrite derived from NPGenerator V2.0.2.
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

import { BlockConstraint, sqrt } from "./core.js";
import { readBlockArray } from "./grid.js";
import { JavaRandom } from "./random.js";
import type { NumberPlaceFile } from "./xml.js";

export const DEFAULT_SIZE = 9;
export const MIN_SIZE = 2;
export const MAX_SIZE = 25;

export interface Variant {
  size: number;
  block: BlockConstraint;
  blockArray: number[];
  vertical: boolean;
  horizontal: boolean;
  diagonal: boolean;
  defaultBlock: boolean;
}

class ProblemBuilder {
  vertical = true;
  horizontal = true;
  diagonal = false;
  defaultBlock = true;
  rectangleWidth = -1;
  rectangleHeight = -1;
  readonly groups: number[][] = [];

  constructor(readonly size: number) {}

  addGroup(labels: number[]): void {
    this.groups.push(labels);
  }

  build(diagonalLast = false): number[][] {
    const blocks: number[][] = [];
    if (this.vertical) addVertical(blocks, this.size);
    if (this.horizontal) addHorizontal(blocks, this.size);
    if (this.diagonal && !diagonalLast) addDiagonal(blocks, this.size);
    if (this.defaultBlock) {
      const square = sqrt(this.size);
      addRectangle(blocks, this.size, square, square);
    } else if (this.rectangleWidth > 0 && this.rectangleHeight > 0) {
      addRectangle(
        blocks,
        this.size,
        this.rectangleWidth,
        this.rectangleHeight,
      );
    }
    for (const labels of this.groups) addByArray(blocks, labels, this.size);
    if (this.diagonal && diagonalLast) addDiagonal(blocks, this.size);
    return blocks;
  }
}

function addVertical(blocks: number[][], size: number): void {
  for (let column = 0; column < size; column++) {
    blocks.push(Array.from({ length: size }, (_, row) => row * size + column));
  }
}

function addHorizontal(blocks: number[][], size: number): void {
  for (let row = 0; row < size; row++) {
    blocks.push(Array.from({ length: size }, (_, column) => row * size + column));
  }
}

function addDiagonal(blocks: number[][], size: number): void {
  blocks.push(Array.from({ length: size }, (_, index) => index * size + index));
  blocks.push(
    Array.from(
      { length: size },
      (_, index) => (size - 1 - index) * size + index,
    ),
  );
}

function addRectangle(
  blocks: number[][],
  size: number,
  width: number,
  height: number,
): void {
  for (let column = 0; column < size; column += width) {
    for (let row = 0; row < size; row += height) {
      const block: number[] = [];
      for (let x = 0; x < width; x++) {
        for (let y = 0; y < height; y++) {
          block.push((row + y) * size + column + x);
        }
      }
      blocks.push(block);
    }
  }
}

function addByArray(blocks: number[][], labels: number[], size: number): void {
  const labelToIndex = new Map<number, number>();
  for (const label of labels) {
    if (label !== 0 && !labelToIndex.has(label)) {
      labelToIndex.set(label, labelToIndex.size);
    }
  }
  const groups = Array.from({ length: labelToIndex.size }, () => [] as number[]);
  for (let cell = 0; cell < labels.length; cell++) {
    const label = labels[cell]!;
    if (label !== 0) groups[labelToIndex.get(label)!]!.push(cell);
  }
  for (const group of groups) {
    if (group.length !== size) {
      throw new Error(`every block must contain exactly ${size} cells`);
    }
    blocks.push(group);
  }
}

class DisjointSet {
  private readonly parent: number[];

  constructor(size: number) {
    this.parent = Array.from({ length: size }, (_, index) => index);
  }

  find(value: number): number {
    const parent = this.parent[value]!;
    if (parent === value) return value;
    const root = this.find(parent);
    this.parent[value] = root;
    return root;
  }

  union(first: number, second: number): void {
    this.parent[this.find(first)] = this.find(second);
  }

  isSameGroup(first: number, second: number): boolean {
    return this.find(first) === this.find(second);
  }
}

/** Exact port of the original core/BlockSplit traversal and random draws. */
class BlockSplit {
  private verticalWall: boolean[][] = [];
  private horizontalWall: boolean[][] = [];

  constructor(
    private readonly width: number,
    private readonly height: number,
    private readonly random: JavaRandom,
  ) {}

  private index(x: number, y: number, width: number): number {
    return y * width + x;
  }

  private generateMaze(width: number, height: number): void {
    this.verticalWall = Array.from(
      { length: width + 1 },
      () => Array(height + 1).fill(false) as boolean[],
    );
    this.horizontalWall = Array.from(
      { length: width + 1 },
      () => Array(height + 1).fill(false) as boolean[],
    );
    const groups = new DisjointSet(width * height);
    const edges: Array<[number, number]> = [];
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        this.verticalWall[x]![y] = true;
        this.horizontalWall[x]![y] = true;
        const cell = this.index(x, y, width);
        if (x + 1 < width) edges.push([cell, this.index(x + 1, y, width)]);
        if (y + 1 < height) edges.push([cell, this.index(x, y + 1, width)]);
      }
    }
    this.random.shuffle(edges);
    let remaining = width * height - 1;
    for (const [first, second] of edges) {
      if (remaining <= 0) break;
      const firstX = first % width;
      const firstY = Math.trunc(first / width);
      const secondX = second % width;
      const secondY = Math.trunc(second / width);
      if (groups.isSameGroup(first, second)) continue;
      groups.union(first, second);
      if (firstX === secondX) {
        this.verticalWall[firstX]![firstY] = false;
      } else if (firstY === secondY) {
        this.horizontalWall[firstX]![firstY] = false;
      }
      remaining--;
    }
  }

  private isMovable(px: number, py: number, qx: number, qy: number): boolean {
    if (px === qx) {
      if (py > qy) return this.isMovable(px, qy, qx, py);
      if (py + 1 !== qy) return false;
      if ((py & 1) === 0) return true;
      return !this.verticalWall[Math.trunc(px / 2)]![Math.trunc(py / 2)];
    }
    if (py === qy) {
      if (px > qx) return this.isMovable(qx, py, px, qy);
      if (px + 1 !== qx) return false;
      if ((px & 1) === 0) return true;
      return !this.horizontalWall[Math.trunc(px / 2)]![Math.trunc(py / 2)];
    }
    return false;
  }

  private walkMaze(): number[] {
    const dx = [0, 1, 0, -1];
    const dy = [1, 0, -1, 0];
    let x = this.random.nextInt(Math.trunc(this.width / 2)) * 2;
    let y = this.random.nextInt(Math.trunc(this.height / 2)) * 2;
    let direction = 0;
    const result = Array(this.width * this.height).fill(-1) as number[];
    let id = 0;
    let count = 0;
    let yet = true;
    while (yet) {
      yet = false;
      result[this.index(x, y, this.width)] = id;
      count++;
      if (count === this.width) {
        id++;
        count = 0;
      }
      for (let index = 3; index < 7; index++) {
        const nextDirection = (index + direction) % 4;
        const nextX = x + dx[nextDirection]!;
        const nextY = y + dy[nextDirection]!;
        if (
          this.width % 2 !== 0 &&
          this.height % 2 !== 0 &&
          nextX === this.width - 1 &&
          nextY === this.height - 1 &&
          count !== 0 &&
          result[this.index(nextX, nextX, this.width)] === -1
        ) {
          // The original intentionally uses nextX for both coordinates here.
          result[this.index(nextX, nextX, this.width)] = id;
          count++;
          if (count === this.width) {
            id++;
            count = 0;
          }
        }
        if (
          nextX >= 0 && nextX < this.width &&
          nextY >= 0 && nextY < this.height &&
          this.isMovable(x, y, nextX, nextY) &&
          result[this.index(nextX, nextY, this.width)] === -1
        ) {
          yet = true;
          direction = nextDirection;
          x = nextX;
          y = nextY;
          break;
        }
      }
    }
    return result;
  }

  splitBlock(): number[] {
    const width = Math.trunc(this.width / 2);
    const height = Math.trunc(this.height / 2);
    this.generateMaze(width, height);
    if (this.width % 2 !== 0) {
      for (let index = 0; index < height; index++) {
        this.verticalWall[width]![index] = true;
        this.horizontalWall[width - 1]![index] = false;
      }
    }
    if (this.height % 2 !== 0) {
      for (let index = 0; index < width; index++) {
        this.verticalWall[index]![height - 1] = false;
        this.horizontalWall[index]![height] = true;
      }
    }
    return this.walkMaze();
  }
}

export function rectangleBlockArray(
  size: number,
  width: number,
  height: number,
): number[] {
  const labels = Array(size * size).fill(0) as number[];
  const blocksAcross = Math.trunc(size / width);
  for (let row = 0; row < size; row++) {
    for (let column = 0; column < size; column++) {
      labels[row * size + column] =
        Math.trunc(row / height) * blocksAcross + Math.trunc(column / width) + 1;
    }
  }
  return labels;
}

function normalizeBlockArray(size: number, labels: number[]): number[] {
  if (labels.length !== size * size) {
    throw new Error(`block grid must contain exactly ${size * size} cells`);
  }
  const normalizedLabels = new Map<number, number>();
  const result: number[] = [];
  const counts = Array(size).fill(0) as number[];
  for (const label of labels) {
    let normalized = normalizedLabels.get(label);
    if (normalized === undefined) {
      if (normalizedLabels.size === size) {
        throw new Error(`block grid must contain exactly ${size} blocks`);
      }
      normalized = normalizedLabels.size + 1;
      normalizedLabels.set(label, normalized);
    }
    result.push(normalized);
    counts[normalized - 1]!++;
  }
  if (normalizedLabels.size !== size) {
    throw new Error(`block grid must contain exactly ${size} blocks`);
  }
  if (counts.some((count) => count !== size)) {
    throw new Error(`every block must contain exactly ${size} cells`);
  }
  return result;
}

function finishVariant(
  builder: ProblemBuilder,
  blockArray: number[],
  defaultBlock: boolean,
  xmlOrder: boolean,
): Variant {
  return {
    size: builder.size,
    block: new BlockConstraint(builder.build(xmlOrder), builder.size),
    blockArray: [...blockArray],
    vertical: builder.vertical,
    horizontal: builder.horizontal,
    diagonal: builder.diagonal,
    defaultBlock,
  };
}

function baseBuilder(
  size: number,
  vertical: boolean,
  horizontal: boolean,
  diagonal: boolean,
): ProblemBuilder {
  const builder = new ProblemBuilder(size);
  builder.vertical = vertical;
  builder.horizontal = horizontal;
  builder.diagonal = diagonal;
  return builder;
}

function buildArrayVariant(
  size: number,
  labels: number[],
  vertical: boolean,
  horizontal: boolean,
  diagonal: boolean,
  xmlOrder: boolean,
): Variant {
  const normalized = normalizeBlockArray(size, labels);
  const builder = baseBuilder(size, vertical, horizontal, diagonal);
  builder.defaultBlock = false;
  builder.addGroup(normalized);
  return finishVariant(builder, normalized, false, xmlOrder);
}

export async function buildVariant(
  size: number,
  blockSpec: string | undefined,
  vertical: boolean,
  horizontal: boolean,
  diagonal: boolean,
  random: JavaRandom,
  xmlOrder: boolean,
): Promise<Variant> {
  requireSize(size);
  if (blockSpec === undefined) {
    const square = sqrt(size);
    if (square * square !== size) {
      throw new Error("--blocks is required when --size is not a perfect square");
    }
    const builder = baseBuilder(size, vertical, horizontal, diagonal);
    builder.defaultBlock = true;
    return finishVariant(
      builder,
      rectangleBlockArray(size, square, square),
      true,
      xmlOrder,
    );
  }
  if (blockSpec === "random") {
    return buildArrayVariant(
      size,
      new BlockSplit(size, size, random).splitBlock(),
      vertical,
      horizontal,
      diagonal,
      xmlOrder,
    );
  }
  if (blockSpec.startsWith("@")) {
    if (blockSpec.length === 1) {
      throw new Error("--blocks @file requires a file name");
    }
    return buildArrayVariant(
      size,
      await readBlockArray(blockSpec.slice(1), size),
      vertical,
      horizontal,
      diagonal,
      xmlOrder,
    );
  }
  const dimensions = blockSpec.toLowerCase().split("x");
  if (dimensions.length !== 2) {
    throw new Error("--blocks must be WxH, random, or @file.txt");
  }
  if (
    !/^[+-]?\d+$/.test(dimensions[0]!) ||
    !/^[+-]?\d+$/.test(dimensions[1]!)
  ) {
    throw new Error("--blocks WxH requires integer dimensions");
  }
  const width = Number(dimensions[0]);
  const height = Number(dimensions[1]);
  if (
    !Number.isSafeInteger(width) ||
    !Number.isSafeInteger(height) ||
    width <= 0 ||
    height <= 0 ||
    width * height !== size ||
    size % width !== 0 ||
    size % height !== 0
  ) {
    throw new Error("--blocks WxH requires W*H == size");
  }
  const builder = baseBuilder(size, vertical, horizontal, diagonal);
  builder.defaultBlock = false;
  builder.rectangleWidth = width;
  builder.rectangleHeight = height;
  return finishVariant(
    builder,
    rectangleBlockArray(size, width, height),
    false,
    xmlOrder,
  );
}

export function buildXmlVariant(
  source: NumberPlaceFile,
  forceDiagonal: boolean,
  noVertical: boolean,
  noHorizontal: boolean,
): Variant {
  const size = source.numSize;
  requireSize(size);
  const vertical = source.vertical && !noVertical;
  const horizontal = source.horizontal && !noHorizontal;
  const diagonal = source.diagonal || forceDiagonal;
  const builder = baseBuilder(size, vertical, horizontal, diagonal);
  let blockArray: number[];
  if (source.defaultBlock) {
    const square = sqrt(size);
    if (square * square !== size) {
      throw new Error("XML default-block requires a perfect-square size");
    }
    builder.defaultBlock = true;
    blockArray = rectangleBlockArray(size, square, square);
  } else {
    if (source.groupArrays.length === 0) {
      throw new Error("XML custom block constraint is missing <group>");
    }
    builder.defaultBlock = false;
    blockArray = [...source.groupArrays[0]!];
  }
  for (const group of source.groupArrays) builder.addGroup(group);
  return finishVariant(builder, blockArray, source.defaultBlock, true);
}

export function requireSize(size: number): void {
  if (size < MIN_SIZE || size > MAX_SIZE) {
    throw new Error("--size must be between 2 and 25");
  }
}
